### PR TITLE
fix: tooltips not working as expected

### PR DESCRIPTION
**Pull request by pac-guerreiro**
### Summary

This PR fixes issues related to the tooltip component not working correctly on web and native, as reported on #3729.

The changes are as follows:

- Fixed console warnings when using tooltip component with components that didn't forward refs. I introduced changes to how the tooltip component handles hovering/pressing events, so now there's no need to pass refs to wrapped components.

- Fixed tooltip not working with non-pressable components, like avatar text, on both web and native.

- Fixed tooltip not working with FAB component. First part of the problem was that FAB didn't use handlers for hover/press events, so the tooltip would not be triggered to show. Second part of the problem was that when using tooltip with a component that has a `position: 'absolute'` the tooltip isn't placed correctly on the screen. The solution was to take the component styling into account when calculating the right placement for the tooltip.

- Added reproducibility sample from #3729 to Tooltips screen/page on the example app.

- Changed how tooltips are triggered:
  - on web, the tooltip is triggered if the user mouse pointer hovers the wrapped component;
  - on native, the tooltip is triggered by long pressing the wrapped component.

### Test plan

These changes can be tested on the Tooltips screen/page of the example app.

**Screen recordings**

Web:

https://user-images.githubusercontent.com/48553277/226315643-f7062a28-dd82-4a64-8870-40ca1ebdeae9.mov

---

Native:

https://user-images.githubusercontent.com/48553277/226315654-428def6c-6544-4b52-9621-b8f4f794650f.mov

### CODE REVIEW

File: src/react-navigation/useInternalNavigation.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { useNavigation, NavigationProp } from '@react-navigation/native';
+
+import type { NavigationParams } from './types';
+
+/**
+ * Get properly typed navigation object from useNavigation hook
+ *
+ * @returns navigation object
+ */
+const useInternalNavigation = (): NavigationProp<NavigationParams> =>
```
Comment (lukewalczak) [inline src/react-navigation/useInternalNavigation.tsx:10]: Similarly to my previous comment, please notice that Paper is independent of `react-navigation` – we are using it in the example app, but not in the library itself.

**Comment by tjaniczek** [comment]
Any chance we can implement immediate tooltip hiding on something like `onMouseLeave` on web, to avoid this situation?

image

**Comment by pac-guerreiro** [comment]
> Any chance we can implement immediate tooltip hiding on something like `onMouseLeave` on web, to avoid this situation?
> 
> 
> 
> image
> 
> 

There's already a 'leaveTouchDelay' prop but it has a default value of 1500. I can change the value to 0 when Platform.OS === 'web', what do you think?

**Comment by DimitarNestorov** [comment]
> There's already a 'leaveTouchDelay' prop but it has a default value of 1500. I can change the value to 0 when Platform.OS === 'web', what do you think?

`Platform.OS === 'web'` doesn't necessarily mean that the user has a mouse pointer, you could be browsing on your smartphone. `Platform.OS === 'ios' || Platform.OS === 'android'` also doesn't necessarily mean touch, you may have a mouse connected. Both hover and touch tooltips should be available on all platforms. I say we don't modify the current behaviour but definitely change this with Paper 6.0.
[Pressable has `onHoverIn` and `onHoverOut` since RN 0.68](https://reactnative.dev/docs/0.68/pressable#onhoverin).
[Pointer events are available in the new architecture since RN 0.71](https://reactnative.dev/blog/2023/01/12/version-071#events).
`leaveTouchDelay` is a bad name for the prop. Maybe we can rename it to `delayPressOut` ([inspired by `TouchableWithoutHighlight` which has had a prop with the same name since RN 0.60 at least](https://reactnative.dev/docs/0.60/touchablewithoutfeedback#delaypressout)).
And also a separate `delayHoverOut`/`delayPointerLeave` would be nice, that one can be `0` by default.

File: src/components/Tooltip/Tooltip.tsx [hunk 2]
```diff
@@ -193,17 +209,16 @@ const Tooltip = ({
...
+      <Pressable
+        ref={childrenWrapperRef}
+        style={isWeb ? styles.webContainer : null}
+        {...(isWeb ? webPressProps : mobilePressProps)}
       >
         {React.cloneElement(children, {
           ...rest,
-          ref: childrenWrapperRef,
           ...(isWeb ? webPressProps : mobilePressProps),
         })}
```
Comment (DimitarNestorov) [inline src/components/Tooltip/Tooltip.tsx:220]:
```diff
-         {React.cloneElement(children, {
-            ...rest,
-            ref: childrenWrapperRef,
-            ...(isWeb ? webPressProps : mobilePressProps),
-          })}
+          {children}
```
If the container has the press handlers the child doesn't need them.

**Review by DimitarNestorov — CHANGES_REQUESTED** [review]

File: src/components/Tooltip/Tooltip.tsx (hunk 2 above)
Comment (pac-guerreiro) [reply src/components/Tooltip/Tooltip.tsx:220]: In case the child is a button, it is needed, otherwise the tooltip wont be triggered

**Comment by pac-guerreiro** [comment]
> > There's already a 'leaveTouchDelay' prop but it has a default value of 1500. I can change the value to 0 when Platform.OS === 'web', what do you think?
> 
> `Platform.OS === 'web'` doesn't necessarily mean that the user has a mouse pointer, you could be browsing on your smartphone. `Platform.OS === 'ios' || Platform.OS === 'android'` also doesn't necessarily mean touch, you may have a mouse connected. Both hover and touch tooltips should be available on all platforms. I say we don't modify the current behaviour but definitely change this with Paper 6.0. [Pressable has `onHoverIn` and `onHoverOut` since RN 0.68](https://reactnative.dev/docs/0.68/pressable#onhoverin). [Pointer events are available in the new architecture since RN 0.71](https://reactnative.dev/blog/2023/01/12/version-071#events). `leaveTouchDelay` is a bad name for the prop. Maybe we can rename it to `delayPressOut` ([inspired by `TouchableWithoutHighlight` which has had a prop with the same name since RN 0.60 at least](https://reactnative.dev/docs/0.60/touchablewithoutfeedback#delaypressout)). And also a separate `delayHoverOut`/`delayPointerLeave` would be nice, that one can be `0` by default.

Wow these are very good valid points! Thank you for mentioning these @DimitarNestorov!

I can apply these changes you suggested 😄

**Comment by tjaniczek** [comment]
Closes https://github.com/callstack/react-native-paper/issues/3729

**Review by DimitarNestorov — APPROVED** [review]

File: src/components/__tests__/Tooltip.test.tsx [hunk 3]
```diff
@@ -65,204 +65,389 @@ describe('Tooltip', () => {
     return { wrapper };
   };
 
-  describe('Unmount', () => {
-    beforeAll(() => jest.spyOn(global, 'clearTimeout'));
-    afterEach(() => jest.clearAllMocks());
+  describe('Mobile', () => {
```
[the remark this answers is not in the document]
Comment (tjaniczek) [reply src/components/__tests__/Tooltip.test.tsx:68]: Yeah absolute positioning is still a bit wonky at the moment. I think in the interest of time we'll merge this piece now and create an issue to rework the positioning. We can address the tests files as well there.

File: src/components/FAB/FAB.tsx [hunk 4]
```diff
@@ -286,6 +286,7 @@ const FAB = forwardRef<View, Props>(
           accessibilityState={newAccessibilityState}
           testID={testID}
           style={{ borderRadius }}
+          {...rest}
```
[the remark this answers is not in the document]
Comment (tjaniczek) [reply src/components/FAB/FAB.tsx:289]: @pac-guerreiro remember why it was added there?

File: src/components/Tooltip/Tooltip.tsx [hunk 5]
```diff
@@ -125,10 +125,15 @@ const Tooltip = ({
       clearTimeout(hideTooltipTimer.current);
     }
 
-    showTooltipTimer.current = setTimeout(() => {
+    if (isWeb) {
+      showTooltipTimer.current = setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (tjaniczek) [reply src/components/Tooltip/Tooltip.tsx:129]: Changing prop would be a breaking change and require releasing new version. `enterTouchDelay` is translated to `delayLongPress` on mobile: https://github.com/callstack/react-native-paper/blob/39f257b43d05b49c8cbe7f75c7bd6540f64900b0/src/components/Tooltip/Tooltip.tsx#L161

**Pull request merged**